earthfile2llb/wait_block.go

```diff
@@ -0,0 +1,194 @@
+			metadata[refPrefix+"/insecure-push"] = []byte("true")
+		}
+		metadata[refPrefix+"/"+exptypes.ExporterImageConfigKey] = config
+		refID++
+
+		if isMultiPlatform[saveImage.DockerTag] {
+			platformStr := saveImage.Platform.String()
+			platformImgName, err := llbutil.PlatformSpecificImageName(saveImage.DockerTag, saveImage.Platform)
+			if err != nil {
+				return err
+			}
+
+			if saveImage.CheckDuplicate && saveImage.DockerTag != "" {
+				if _, found := platformImgNames[platformImgName]; found {
+					return errors.Errorf(
+						"image %s is defined multiple times for the same platform (%s)",
+						saveImage.DockerTag, platformImgName)
+				}
+				platformImgNames[platformImgName] = true
+			}
+
+			metadata[refPrefix+"/platform"] = []byte(platformStr)
+		}
+	}
+
+	if len(w.converters) == 0 {
+		panic("saveImagesWaitItem should never have been created with zero converters")
+	}
+	gatewayClient := w.converters[0].opt.GwClient // could be any converter's gwClient (they should app be the same)
+
+	err := gatewayClient.SaveImage(ctx, gwclient.SaveImageRequest{
+		Refs:     refs,
+		Metadata: metadata,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to SAVE IMAGE")
+	}
+	return nil
+}
+
+func (w *runCommandWaitItem) wait(ctx context.Context) error {
+	state := *w.cmd
+	ref, err := llbutil.StateToRef(
+		ctx, w.c.opt.GwClient, state, w.c.opt.NoCache,
+		w.c.platr, w.c.opt.CacheImports.AsMap())
+	if err != nil {
+		return errors.Wrap(err, "waiting for RUN state to ref")
+	}
+
+	// wait for ref to be solved
+	_, err = ref.ReadDir(ctx, gwclient.ReadDirRequest{Path: "/"})
+	if err != nil {
+		return errors.Wrap(err, "waiting for RUN command to complete")
+	}
+	return nil
+}
+
+type waitBlock struct {
+	items []waitItem
+	mu    sync.Mutex
+}
+
+func newWaitBlock() *waitBlock {
+	return &waitBlock{}
+}
+
+func (wb *waitBlock) lastItemUnsafe() (waitItem, bool) {
+	n := len(wb.items)
+	if n > 0 {
+		return wb.items[n-1], true
+	}
+	return nil, false
+}
+
+func (wb *waitBlock) getOrCreateSaveImageWaitItemUnsafe() *saveImagesWaitItem {
+	var siwi *saveImagesWaitItem
+	item, ok := wb.lastItemUnsafe()
+	if ok {
+		siwi, ok = item.(*saveImagesWaitItem)
+	}
+	if !ok {
+		siwi = &saveImagesWaitItem{}
+		wb.items = append(wb.items, siwi)
+	}
+	return siwi
+}
+
+func (wb *waitBlock) addSaveImage(si states.SaveImage, c *Converter, push bool) {
+	wb.mu.Lock()
+	defer wb.mu.Unlock()
+	item := wb.getOrCreateSaveImageWaitItemUnsafe()
+	item.converters = append(item.converters, c)
+	item.images = append(item.images, si)
+	item.push = append(item.push, push)
+}
+
+func (wb *waitBlock) addCommand(cmd *pllb.State, c *Converter) {
+	wb.mu.Lock()
+	defer wb.mu.Unlock()
+	item := runCommandWaitItem{
+		c:   c,
+		cmd: cmd,
+	}
+	wb.items = append(wb.items, &item)
+}
+
+func (wb *waitBlock) wait(ctx context.Context) error {
+	wb.mu.Lock()
+	defer wb.mu.Unlock()
+	for _, item := range wb.items {
+		err := item.wait(ctx)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
I think these items could be fired off in parallel - otherwise, the waiting is relatively slow as this only progress one by one.

There is an `ErrorGroup` in the converter opts which we could use. And also, we should probably use a semaphore to limit parallelism. This kind of semaphore could be appropriate here too: https://github.com/earthly/earthly/blob/main/earthfile2llb/withdockerrun.go#L72-L74.

Reply:
That would break the use-case for this feature: ensuring all `SAVE IMAGE --push` commands finish before performing a `RUN ./deploy` command.